include/ringbuffer_utils.hpp

```diff
@@ -152,6 +153,151 @@ class PerfRingBufferReader {
   uint64_t _head;
 };
 
+struct MPSCRingBufferHeader {
+  uint64_t size;
+  static constexpr uint64_t k_discard_bit = 1UL << 62;
+  static constexpr uint64_t k_busy_bit = 1UL << 63;
+
+  static bool is_busy(uint64_t size) { return size & k_busy_bit; }
+  static bool is_discarded(uint64_t size) { return size & k_discard_bit; }
+};
+
+class MPSCRingBufferWriter {
+public:
+  explicit MPSCRingBufferWriter(RingBuffer &rb) : _rb(rb) {
+    assert(rb.type == RingBufferType::kMPSCRingBuffer);
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+  }
+
+  MPSCRingBufferWriter(const MPSCRingBufferWriter &) = delete;
+  MPSCRingBufferWriter &operator=(const MPSCRingBufferWriter &) = delete;
+
+  void update_tail() {
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+  }
+
+  Buffer reserve(size_t n) {
+    size_t n2 = align_up(n + sizeof(MPSCRingBufferHeader), 8);
+    if (n2 == 0) {
+      return {};
+    }
+
+    // \fixme{nsavoire} Not sure if spinlock is the best option here
+    lock(reinterpret_cast<std::atomic_flag *>(_rb.spinlock));
+
+    // No need for atomic operation, since we hold the lock
+    uint64_t writer_pos = *_rb.writer_pos;
+
+    uint64_t new_writer_pos = writer_pos + n2;
+    // Check that there is enough free space
+    if (_rb.mask < new_writer_pos - _tail) {
+      unlock(reinterpret_cast<std::atomic_flag *>(_rb.spinlock));
+      return {};
+    }
+
+    uint64_t head_linear = writer_pos & _rb.mask;
+    MPSCRingBufferHeader *hdr =
+        reinterpret_cast<MPSCRingBufferHeader *>(_rb.data + head_linear);
+
+    // Mark the sample as busy
+    hdr->size = n | MPSCRingBufferHeader::k_busy_bit;
+
+    // Atomic operation required to synchronize with reader load_acquire
```

Comment:
Right!  Even though we're locked, we need to push down the write to the _next_ reader, in case the write doesn't go through before we unlock.

Reply:
Indeed the lock is only used by writers, reader/writer synchronization relies on atomic loads/stores.